osu.Game/Database/EFToRealmMigrator.cs

```diff
@@ -4,52 +4,130 @@
+
+        [Resolved]
+        private RealmContextFactory realmContextFactory { get; set; } = null!;
+
+        [Resolved]
+        private OsuConfigManager config { get; set; } = null!;
+
+        [Resolved]
+        private Storage storage { get; set; } = null!;
+
+        private readonly OsuSpriteText currentOperationText;
+
+        public EFToRealmMigrator()
         {
-            this.efContextFactory = efContextFactory;
-            this.realmContextFactory = realmContextFactory;
-            this.config = config;
-            this.storage = storage;
+            RelativeSizeAxes = Axes.Both;
+
+            InternalChildren = new Drawable[]
+            {
+                new FillFlowContainer
+                {
+                    AutoSizeAxes = Axes.Both,
+                    Direction = FillDirection.Vertical,
+                    Anchor = Anchor.Centre,
+                    Origin = Anchor.Centre,
+                    Spacing = new Vector2(10),
+                    Children = new Drawable[]
+                    {
+                        new OsuSpriteText
+                        {
+                            Anchor = Anchor.Centre,
+                            Origin = Anchor.Centre,
+                            Text = "Database migration in progress",
+                            Font = OsuFont.Default.With(size: 40)
+                        },
+                        new OsuSpriteText
+                        {
+                            Anchor = Anchor.Centre,
+                            Origin = Anchor.Centre,
+                            Text = "This could take a few minutes depending on the speed of your disk(s).",
+                            Font = OsuFont.Default.With(size: 30)
+                        },
+                        new OsuSpriteText
+                        {
+                            Anchor = Anchor.Centre,
+                            Origin = Anchor.Centre,
+                            Text = "Please keep the window open until this completes!",
+                            Font = OsuFont.Default.With(size: 30)
+                        },
+                        new LoadingSpinner(true)
+                        {
+                            State = { Value = Visibility.Visible }
+                        },
+                        currentOperationText = new OsuSpriteText
+                        {
+                            Anchor = Anchor.Centre,
+                            Origin = Anchor.Centre,
+                            Font = OsuFont.Default.With(size: 30)
+                        },
+                    }
+                },
+            };
         }
 
-        public void Run()
+        protected override void LoadComplete()
         {
+            base.LoadComplete();
+
+            // needs to be run on the update thread because of realm BlockAllOperations.
+            // maybe we can work around this? not sure..
```

Comment:
Is this comment still valid? this is invoked in `LoadComplete` which should guarantee running from update thread.